pkg/workload/workload_generator/sql_generator.go

```diff
@@ -0,0 +1,257 @@
+//     a) Open “crdb_internal.node_statement_statistics.txt” and scan its TSV contents.
+//     b) Read the header row to map column names to indices.
+//     c) For each data row:
+//     • Filter out rows not matching dbName.
+//     • Skip internal “job id=” entries.
+//     • Extract txn_fingerprint_id and raw SQL (column “key”), dropping statements
+//     containing DEALLOCATE or WHEN.
+//     • Unquote the TSV literal (strip outer quotes, unescape "").
+//     • Call replacePlaceholders(rawSQL, allSchemas) to substitute schema-driven placeholders.
+//     • Append the rewritten statement to txnMap[txnID], and record txnID in txnOrder if first-seen.
+//  3. After all nodes are processed, determine output file paths:
+//     sqlLocation/<dbName>_read.sql and sqlLocation/<dbName>_write.sql.
+//  4. Create or truncate these two files, then defer their closure.
+//  5. Call writeTransaction(txnOrder, txnMap, outReadFile, outWriteFile) to emit the workloads.
+//  6. Propagate any I/O, scanning, or SQL-rewriting errors.
+//
+// Returns:
+//
+//	error – if any file I/O, scanner error, or placeholder-replacement error occurs.
+func generateWorkload(
+	debugLogs string, allSchemas map[string]*TableSchema, dbName, sqlLocation string,
+) error {
+	// 1) Grouping structures are prepared.
+	txnOrder := make([]string, 0)       // first-seen txn IDs
+	txnMap := make(map[string][]string) // txnID → []SQL statements
+
+	// 2) Iterating node directories (nodes/1, nodes/2, …).
+	nodesRoot := filepath.Join(debugLogs, "nodes")
+	for node := 1; ; node++ {
+		nodeDir := filepath.Join(nodesRoot, strconv.Itoa(node))
+		if fi, err := os.Stat(nodeDir); err != nil || !fi.IsDir() {
+			break
+		}
+
+		// 3) The statistics tsv is opened to read the transactions from the debug logs.
+		statsPath := filepath.Join(nodeDir, "crdb_internal.node_statement_statistics.txt")
+		f, err := os.Open(statsPath)
+		if err != nil {
+			return errors.Wrapf(err, "opening %s", statsPath)
+		}
+		scanner := bufio.NewScanner(f)
+
+		// 4) Header row and map column→index are read.
+		if !scanner.Scan() {
+			if err := f.Close(); err != nil {
+				return err
+			}
+			continue
+		}
+		columnIndex, err := getColumnIndexes(scanner, f, statsPath)
+		if err != nil {
+			return err
+		}
+
+		// 5) Each data row is scanned.
+		for scanner.Scan() {
+			row := strings.Split(scanner.Text(), "\t")
+			// 5a) Filtering by database_name.
+			if row[columnIndex["database_name"]] != dbName {
+				continue
+			}
+			// 5b) Internal “job id=” lines are skipped.
+			app := row[columnIndex["application_name"]]
+			if strings.Contains(app, "job id=") {
+				continue
+			}
+			// 5c) txnID and raw SQL are extracted.
+			txnID := row[columnIndex["txn_fingerprint_id"]]
+			rawSQL := row[columnIndex["key"]]
+			// These are unhandled clauses for now.
+			var skipRE = regexp.MustCompile(`\b(DEALLOCATE|WHEN)\b`)
+			if skipRE.MatchString(rawSQL) {
+				continue // dropping any statement containing DEALLOCATE or WHEN
```

Comment:
Can we move them to the unhandled bucket?

Reply:
I believe it makes sense to move any transaction with even a single unhandled query to the unhandled file. The current behavior is breaking atomicity, which could lead to adverse effects. Could you please let me know how many transactions will move to the unhandled file after this?